coffeecard/CoffeeCard.Library/Services/v2/TicketService.cs

```diff
@@ -49,5 +55,46 @@ public Task<List<TicketResponse>> GetTickets(User user, bool includeUsed)
                     ProductName = t.Purchase.ProductName
                 }).ToListAsync();
         }
+        
+        public async Task<TicketDto> UseTicket(int userId, int productId)
+        {
+            Log.Information($"Using ticket with id, {productId}");
+            var ticket = GetFirstTicketFromProduct(productId, userId);
+
+            ticket.IsUsed = true;
+            ticket.DateUsed = DateTime.UtcNow;
+            
+            if (ticket.Purchase.Price != 0) //Paid products increases your rank on the leaderboard
+            {
+                await _statisticService.IncreaseStatisticsBy(userId, 1);
+            }
+
+            await _context.SaveChangesAsync();
+
+            return new TicketDto
+            {
+                Id = ticket.Id,
+                DateCreated = ticket.DateCreated,
+                DateUsed = ticket.DateUsed,
+                ProductName = ticket.Purchase.ProductName
+            };
+        }
+        
+        private Ticket GetFirstTicketFromProduct(int productId, int userId)
+        {
+            var ticket = _context.Tickets
+                .Include(t => t.Purchase)
+                .FirstOrDefault(t => t.Owner.Id == userId && t.ProductId == productId && !t.IsUsed);
+            
+            if (ticket == null)
+            {
+                throw new ApiException("No tickets found for the given product with this user", StatusCodes.Status404NotFound);
```

Comment:
In API v2, we try to use more specialized exceptions rather than a catch-all ApiException.

Reply:
Since our exception filter is not setup to handle other types of exceptions, this seems undesirable. You will be throwing responses with status 500 to the user, or am I missing something obvious?

Reply:
You are exactly right, we will have to expand the exception filter as more exceptions are added 😉

Reply:
Instead of extending the exception filter to handle every single exception, I have made EntityNotFoundException extend ApiException